File: justjavastaff/src/main/java/com/marknkamau/justjavastaff/JustJavaStaffApp.java
package com.marknkamau.justjavastaff;

import android.app.Application;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.marknkamau.justjavastaff.authentication.AuthServiceImpl;
import com.marknkamau.justjavastaff.authentication.AuthenticationService;
import com.marknkamau.justjavastaff.data.network.OrdersRepository;
import com.marknkamau.justjavastaff.data.network.OrdersRepositoryImpl;

import org.jetbrains.annotations.NotNull;

import timber.log.Timber;

public class JustJavaStaffApp extends Application {
    private SharedPreferences preferences;
    private AuthenticationService auth;
    private OrdersRepository ordersRepository;

    @Override
    public void onCreate() {
        super.onCreate();

        if (BuildConfig.DEBUG) {
            Timber.plant(new Timber.DebugTree() {
                @Override
                protected String createStackElementTag(@NotNull StackTraceElement element) {
                    return "Timber" + super.createStackElementTag(element) + "." + element.getMethodName();
                }
            });
        }

        preferences = PreferenceManager.getDefaultSharedPreferences(this);
        auth = new AuthServiceImpl();
        ordersRepository = new OrdersRepositoryImpl();
    }

    public SharedPreferences getPreferences() {
        return preferences;
    }

    public AuthenticationService getAuth() {
        return auth;
    }

    public OrdersRepository getOrdersRepository() {
        return ordersRepository;
    }
}

File: justjavastaff/src/main/java/com/marknkamau/justjavastaff/JustJavaStaffApp.kt
package com.marknkamau.justjavastaff

import android.app.Application
import com.marknjunge.core.data.network.AuthService
import com.marknjunge.core.data.firebase.OrderService
import com.marknjunge.core.data.firebase.PaymentService
import com.marknjunge.core.data.firebase.UserService
import com.marknjunge.core.di.databaseModule

import com.marknkamau.justjavastaff.data.local.SettingsRespository
import com.marknkamau.justjavastaff.di.appModule
import org.koin.android.ext.android.inject
import org.koin.android.ext.koin.androidContext
import org.koin.core.context.startKoin

import timber.log.Timber

class JustJavaStaffApp : Application() {
    val auth: AuthService by inject()
    val settingsRepository: SettingsRespository by inject()
    val colorUtils: ColorUtils by inject()
    val paymentService: PaymentService by inject()
    val orderService:OrderService by inject()
    val userService: UserService by inject()

    override fun onCreate() {
        super.onCreate()

        if (BuildConfig.DEBUG) {
            Timber.plant(object : Timber.DebugTree() {
                override fun createStackElementTag(element: StackTraceElement): String? {
                    return "Timber" + super.createStackElementTag(element) + "." + element.methodName
                }
            })
        }

        startKoin {
            androidContext(this@JustJavaStaffApp)
            modules(listOf(appModule, databaseModule))
        }
    }
}
